src/Arbor/File/Format/Asif/Write.hs

```diff
@@ -244,6 +250,79 @@ utcTimeMicrosSegment f = genericFold BB.int64LE (Known F.TimeMicros64LE) (fromTi
       fromTime :: TY.UTCTime -> Int64
       fromTime = view (TY.posixTime . TY.microseconds)
 
+-- | Creates a lookup segment where index keys are 'Word16'
+-- Missing values are represented by 'maxValue :: Word16'
+--
+-- @
+-- word16LookupSegment name f = lookupSegment name f (Known F.Word16LE) BB.word16LE
+-- @
+word16LookupSegment :: (MonadResource m, Ord b)
+  => T.Text
+  -> (a -> Maybe b)
+  -> FoldM m b [Segment Handle]
+  -> FoldM m a [Segment Handle]
+word16LookupSegment name f = lookupSegment name f (Known F.Word16LE) BB.word16LE
+
+-- | Creates a lookup segment where index keys are 'Word32'
+-- Missing values are represented by 'maxValue :: Word32'
+--
+-- @
+-- word32LookupSegment name f = lookupSegment name f (Known F.Word32LE) BB.word32LE
+-- @
+word32LookupSegment :: (MonadResource m, Ord b)
+  => T.Text
+  -> (a -> Maybe b)
+  -> FoldM m b [Segment Handle]
+  -> FoldM m a [Segment Handle]
+word32LookupSegment name f = lookupSegment name f (Known F.Word32LE) BB.word32LE
+
+-- | Creates a lookup segment where index keys are 'Word64'
+-- Missing values are represented by 'maxValue :: Word64'
+--
+-- @
+-- word64LookupSegment name f = lookupSegment name f (Known F.Word64LE) BB.word64LE
+-- @
+word64LookupSegment :: (MonadResource m, Ord b)
+  => T.Text
+  -> (a -> Maybe b)
+  -> FoldM m b [Segment Handle]
+  -> FoldM m a [Segment Handle]
+word64LookupSegment name f = lookupSegment name f (Known F.Word64LE) BB.word64LE
+
+-- | Creates a lookup segment for every input into a value in an "inner" dictionary segment.
+-- Missing values are represented as 'maxBound' for the key type.
+lookupSegment :: (MonadResource m, Ord b, Num i, Bounded i)
+  => T.Text                     -- ^ Lookup segment name
+  -> (a -> Maybe b)             -- ^ Extract "dictionary" value
+  -> Whatever F.Format          -- ^ Format of lookup segment
+  -> (i -> BB.Builder)          -- ^ Write a lookup value
+  -> FoldM m b [Segment Handle] -- ^ A fold that represents a dictionary segment
+  -> FoldM m a [Segment Handle]
+lookupSegment name f fmt enc inner =
+  FoldM lstep linit lextract <> handlesM (to f . _Just) inner
+  where
+    linit = do
+      (_, _, h) <- openTempFile Nothing (T.unpack name)
+      pure (h, Map.empty, 0)
+
+    lstep (h, m, c) a =
+      case f a of
+        Nothing -> do
+          liftIO $ BB.hPutBuilder h $ enc maxBound
+          pure (h, m, c)
+        Just b -> do
+          let (v, c', m') = updateMap b c m
+          liftIO $ BB.hPutBuilder h $ enc v
+          pure (h, m', c')
+
+    lextract (h, _, _) = pure
+      [ segment h $ metaFilename name <> metaFormat fmt
+      ]
+
+    updateMap k c m =
+      maybe (c, c+1, Map.insert k c m) (, c, m) (Map.lookup k m)
```

Comment:
More type signatures please 😁

Reply:
What do you mean? Adding type signature for `updateMap` in `where`?
Do you think that `(Num a, Ord k) => k -> a -> Map.Map k a -> (a, a, Map.Map k a)` would help? :)

Reply:
Something like that.  Although consider using type variables from `lookupSegment` with scoped type variables.

Reply:
Or names